module/zfs/dbuf.c

```diff
@@ -981,7 +981,7 @@ dbuf_read_done(zio_t *zio, arc_buf_t *buf, void *vdb)
 		db->db_freed_in_flight = FALSE;
 		dbuf_set_data(db, buf);
 		db->db_state = DB_CACHED;
-	} else if (zio == NULL || zio->io_error == 0) {
+	} else if (err == 0 && (zio == NULL || zio->io_error == 0)) {
```

Comment:
Actually, after looking at this do you think it would be better to create a null zio here and return the error in that?